src/t_cose_encrypt_dec.c

```diff
@@ -0,0 +1,513 @@
+    size_t                 cek_len=PSA_CIPHER_ENCRYPT_OUTPUT_MAX_SIZE(T_COSE_ENCRYPTION_MAX_KEY_LENGTH);
+    psa_key_attributes_t   attributes=PSA_KEY_ATTRIBUTES_INIT;
+    psa_key_handle_t       key_handle=0;
+
+    /* Initialize decoder */
+    QCBORDecode_Init(&DC,
+                     (UsefulBufC){cose,cose_len},
+                     QCBOR_DECODE_MODE_NORMAL);
+
+   /* Make sure the first item is a tag */
+    result=QCBORDecode_GetNext(&DC,&Item);
+
+    /* Check whether tag is CBOR_TAG_ENCRYPT or CBOR_TAG_ENCRYPT0 */
+    if (QCBORDecode_IsTagged(&DC,&Item,CBOR_TAG_ENCRYPT)==false &&
+        QCBORDecode_IsTagged(&DC,&Item,CBOR_TAG_ENCRYPT0)==false) {
+        return(EXIT_FAILURE);
+    }
+
+    /* protected header */
+    result=QCBORDecode_GetNext(&DC,&protected_hdr);
+
+    if (result!=QCBOR_SUCCESS) {
+        return(EXIT_FAILURE);
+    }
+  
+    if (protected_hdr.uDataType!=QCBOR_TYPE_BYTE_STRING) {
+        return(EXIT_FAILURE);
+    }
+  
+    /* Re-initialize to parse protected header */
+    QCBORDecode_Init(&DC2,
+                     (UsefulBufC)
+                     {
+                      protected_hdr.val.string.ptr,
+                      protected_hdr.val.string.len
+                     },
+                     QCBOR_DECODE_MODE_NORMAL);
+
+    QCBORDecode_EnterMap(&DC2,NULL);
+
+    QCBORDecode_GetInt64InMapN(&DC2,COSE_HEADER_PARAM_ALG,&algorithm_id);
+
+    QCBORDecode_ExitMap(&DC2);
+
+    result=QCBORDecode_Finish(&DC2);
+
+    if (result!=QCBOR_SUCCESS) {
+        return(EXIT_FAILURE);
+    }
+  
+    /* unprotected header */
+    QCBORDecode_EnterMap(&DC,NULL);
+
+    QCBORDecode_GetByteStringInMapN(&DC,COSE_HEADER_PARAM_IV,&nonce_cbor);
+
+    if (QCBORDecode_GetError(&DC)!=0) {
+         return(EXIT_FAILURE);
+    }
+  
+    if (me->key_distribution==T_COSE_KEY_DISTRIBUTION_DIRECT) {
+        QCBORDecode_GetByteStringInMapN(&DC,COSE_HEADER_PARAM_KID,&kid_cbor);
+
+        if (QCBORDecode_GetError(&DC)!=0) {
+             return(EXIT_FAILURE);
+        }
+    }
+
+    QCBORDecode_ExitMap(&DC);
+
+    /* Ciphertext */
+    result=QCBORDecode_GetNext(&DC,&Cipher);
+
+    if (result!=QCBOR_SUCCESS) {
+        return(EXIT_FAILURE);
+    }
+  
+    if (Cipher.val.string.len!=0) {
+        ciphertext=(uint8_t *) Cipher.val.string.ptr;
+        ciphertext_len=Cipher.val.string.len;
+        detached_mode=false;
+    } else {
+        ciphertext=detached_ciphertext;
+        ciphertext_len=detached_ciphertext_len;
+        detached_mode=true;
+    }
+
+    /* Two key distribution mechanisms are supported, namely
+     *  - Direct key distribution (where no recipient info is included)
+     *  - HPKE-based key distribution (which requires recipient info)
+     */
+    if (me->key_distribution==T_COSE_KEY_DISTRIBUTION_DIRECT) {
+        if (kid_cbor.len==0 || 
+            strncmp(me->kid.ptr,kid_cbor.ptr,me->kid.len)!=0
+           ) {
+                return( EXIT_FAILURE );
```

Comment:
Also here, spacing should match the rest of the library, in this case no spaces inside the parens.